pkg/forwarder/transaction.go

```diff
@@ -264,41 +325,54 @@ func (t *HTTPTransaction) internalProcess(ctx context.Context, client *http.Clie
 		}
 		codeCount.Add(1)
 		transactionsHTTPErrors.Add(1)
-		tlmTxHTTPErrors.Inc(t.Domain, statusCode)
+		tlmTxHTTPErrors.Inc(t.Domain, transactionEndpointName, statusCode)
 	}
 
 	if resp.StatusCode == 400 || resp.StatusCode == 404 || resp.StatusCode == 413 {
 		log.Errorf("Error code %q received while sending transaction to %q: %s, dropping it", resp.Status, logURL, string(body))
+		transactionsDroppedByEndpoint.Add(transactionEndpointName, 1)
 		transactionsDropped.Add(1)
-		tlmTxDropped.Inc(t.Domain)
+		tlmTxDropped.Inc(t.Domain, transactionEndpointName)
 		return resp.StatusCode, body, nil
 	} else if resp.StatusCode == 403 {
 		log.Errorf("API Key invalid, dropping transaction for %s", logURL)
+		transactionsDroppedByEndpoint.Add(transactionEndpointName, 1)
 		transactionsDropped.Add(1)
-		tlmTxDropped.Inc(t.Domain)
+		tlmTxDropped.Inc(t.Domain, transactionEndpointName)
 		return resp.StatusCode, body, nil
 	} else if resp.StatusCode > 400 {
 		t.ErrorCount++
 		transactionsErrors.Add(1)
-		tlmTxErrors.Inc(t.Domain, "gt_400")
+		tlmTxErrors.Inc(t.Domain, transactionEndpointName, "gt_400")
 		return resp.StatusCode, body, fmt.Errorf("error %q while sending transaction to %q, rescheduling it", resp.Status, logURL)
 	}
 
-	transactionsSuccessful.Add(1)
-	tlmTxSuccess.Inc(t.Domain)
+	tlmTxSuccessCount.Inc(t.Domain, transactionEndpointName)
+	tlmTxSuccessBytes.Add(float64(t.GetPayloadSize()), t.Domain, transactionEndpointName)
+	transactionsSuccessByEndpoint.Add(transactionEndpointName, 1)
+	transactionsSuccessBytes.Add(transactionEndpointName, int64(t.GetPayloadSize()))
+	transactionsSuccess.Add(1)
 
 	loggingFrequency := config.Datadog.GetInt64("logging_frequency")
 
-	if transactionsSuccessful.Value() == 1 {
+	if transactionsSuccess.Value() == 1 {
 		log.Infof("Successfully posted payload to %q, the agent will only log transaction success every %d transactions", logURL, loggingFrequency)
 		log.Tracef("Url: %q payload: %s", logURL, string(body))
 		return resp.StatusCode, body, nil
 	}
-	if transactionsSuccessful.Value()%loggingFrequency == 0 {
+	if transactionsSuccess.Value()%loggingFrequency == 0 {
 		log.Infof("Successfully posted payload to %q", logURL)
 		log.Tracef("Url: %q payload: %s", logURL, string(body))
 		return resp.StatusCode, body, nil
 	}
 	log.Tracef("Successfully posted payload to %q: %s", logURL, string(body))
 	return resp.StatusCode, body, nil
 }
+
+func getTransactionEndpointName(transaction Transaction) string {
+	if transaction != nil {
```

Comment:
is this function ever called with a `nil` transaction?

I assume you really want to check that the endpoint's name is not the empty string. I don't think that ever happens (haven't checked in details though), and even then I think it's OK to return the empty string.

So overall, I think calling `Transaction.GetEndpointName` directly should work fine, and we can get rid of this function. WDYT?

Reply:
It's called with a nil transaction in one of the test cases

Reply:
Is that test case really relevant? (i.e. what's the code that the test case is testing, and why would it try to fetch the endpoint on a nil transaction?)

Reply:
It's this test case. I wasn't quite sure what it's supposed to do.

```go
func TestDomainForwarderSubmitIfStopped(t *testing.T) {
	forwarder := newDomainForwarder("test", 1, 10, 0)

	require.NotNil(t, forwarder)
	assert.NotNil(t, forwarder.sendHTTPTransactions(nil))
}
```

Reply:
what error would `forwarder.sendHTTPTransactions(nil)` return here?

Reply:
It will be a nil pointer panic if I don't check for nil.

Reply:
I meant: what is this test trying to test on the return value of `forwarder.sendHTTPTransactions(nil)`, putting aside the changes in your PR, i.e. why is it trying to check that it returns an error? Basically trying to figure out whether that test makes sense at all. I'm leaning towards saying it doesn't (and `git blame` doesn't provide much info), but let's confirm that.

There is a similar test on the Forwarder here, which makes sense there (https://github.com/DataDog/datadog-agent/blob/808782cc3f63bbcd7291f60065f85893973d789e/pkg/forwarder/forwarder_test.go#L95-L98) since it's there to confirm that a forwarder that's not started return an error when it's used to submit transactions. I'm not sure why we're trying to test something similar on `DomainForwarder`.

Reply:
I'm not sure either what this test is trying to do. I haven't added it and it's not clear to me what it is supposed to do.

Reply:
I could remove it?

Reply:
I checked the error that's returned: it's `the forwarder input queue for test is full: dropping transaction`.
So it doesn't seem like a meaningful test, and in any case this `DomainForwarder.sendHTTPTransactions` is only used by the `Forwarder`, which already returns an error if it's not started. So let's remove this test, and simplify how you're getting the transaction's endpoint name in the code

Reply:
Okay, done.